### PR TITLE
Task 1

### PR DESCRIPTION
**Pull request by Thekote**
# Truque da 11° carta!

No desafio desta semana, utilizando a biblioteca `React` deve ser criado um app que implementa o truque das 11° carta.
Como funciona o truque: https://kipkis.com/Do_the_%2211Th_Card_Trick%22
A estilização é completamente a critério de vocês, o unico requisito é que precisa ser legivel e o usuario entender que aquilo representa uma carta

# Tarefa 1

Criar um deck de 52 cartas, embaralhar e selecionar aleatoriamente 21 cartas para o truque.

### CODE REVIEW

File: src/index.css [hunk 1]
```diff
@@ -1,13 +1,124 @@
-body {
+html,
```
Comment (samflores) [inline src/index.css]: Tente aplicar as metodologia BEM e ITCSS já estudadas.

File: src/utils/cardHandler.js [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    currentDeck[i] = temp;
+  }
+  return currentDeck;
+}
+
+function getFirst21Items(items) {
+  return items.slice(0, 21);
+}
+
+export { cardDeck, shuffleDeck, getFirst21Items };
```
Comment (samflores) [inline src/utils/cardHandler.js]: Evite colocar lógica do problema em arquivos `utils`. Funções utilitárias devem ser funções genéricas que auxiliam outras funções, por exemplo, uma função genérica `shuffle` que recebe um `array` e retorna um `array` embaralhado faz sentido como uma função utilitária. Ela pode ser usada pelo conceito de Deck, mas por qualquer outra parte da aplicação que precise embaralhar um array.

Todos as funções desse arquivo fazem parte de um mesmo conceito de Deck, então faria mais sentido ter uma classe Deck (que é um Model ou Entidade do sitema) que poderia ser usada como:

```js
let deck = new Deck();
deck.shuffle()
let trickCards = deck.getFirst(21);
```

File: src/utils/cardHandler.js [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+    for (let value of VALUES) {
+      deck.push({ suit, value });
+    }
+  }
+  return deck;
+}
+
+const cardDeck = createCardDeck();
+
+function shuffleDeck(deck) {
```
Comment (samflores) [inline src/utils/cardHandler.js]: Você pode escrever uma função bem mais simples para embaralhar um array usando o método `sort` com uma função de comparação que retorne um número aleatório. https://devdocs.io/javascript/global_objects/array/sort

File: src/utils/cardHandler.js [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+  "7",
+  "8",
+  "9",
+  "10",
+  "J",
+  "Q",
+  "K",
+];
+
+function createCardDeck() {
```
Comment (samflores) [inline src/utils/cardHandler.js]: Você poderia escrever essa função usando `flatMap` e `map` ao invés de `for` loops.
https://devdocs.io/javascript/global_objects/array/flatmap

File: src/utils/cardHandler.js [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+  return SUITS.flatMap((suit) => {
+    return VALUES.map((value) => {
+      return { suit, value };
+    });
+  });
+}
```
Comment (samflores) [inline src/utils/cardHandler.js]: Quando uma arrow function tem apenas 1 instrução você pode omitir as chaves e o return:

```js
  return SUITS.flatMap(suit => VALUES.map(value => ({ suit, value })));
```

**Comment by Thekote** [comment]
SPA view on first task
image

File: src/entity/cardDeck.test.js [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+
+describe("cardDeck", () => {
+  it("should return the created deck", () => {
+    expect(cardDeck).toMatchSnapshot();
+  });
+});
+
+describe("shuffleDeck", () => {
+  it("should return a shuffled deck", () => {
+    expect(cardDeck).not.toStrictEqual(shuffleDeck(cardDeck));
```
Comment (samflores) [inline src/entity/cardDeck.test.js:67]: Este teste é não cobre 100% do comportamento esperado de shuffle.
Edge cases:
- se `cardDeck` for um array vazio o teste falha.
- se `cardDeck` possuir apenas um elemento o teste falha
- existe uma chance da função `shuffleDeck` retornar o array na mesma ordem do original, já que ela usa `random`. Quanto menor for o array de entrada, maior essa chance.

File: src/entity/cardDeck.test.js [hunk 7]
```diff
@@ -0,0 +1,28 @@
+import { cardDeck, shuffleDeck } from "./cardDeck";
+import { getArrayItems } from "../utils/arrayItems";
+
+describe("cardDeck", () => {
+  it("should return the created deck", () => {
+    expect(cardDeck).toMatchSnapshot();
```
Comment (samflores) [inline src/entity/cardDeck.test.js]: Eu esqueci de adicionar esse comentário. Foi mals.

Tem alguns problemas nesse teste:

### Testando a coisa errada

Você não está testando a função `createCardDeck`, mas uma constante definida no módulo. Se eu alterar o módulo conforme mostrado abaixo o teste continua passando, mesmo a função `createCardDeck` não funcionando mais:
```js
function createCardDeck() {
  return [];
}
const cardDeck = SUITS.flatMap((suit) => VALUES.map((value) => ({ suit, value })));
```

### Uso de snapshots

Snapshots são indicados para testar a renderização de components, armazenando elementos do DOM criados pelo mesmo.

Eu teria declarado uma constante com o valor esperado no próprio arquivo de teste (opção 1), ou melhor ainda, feito a função `createCardDeck` não depender das constantes `SUITES` e `VALUES` (opção 2) deixando-a mais simples de testar:

opção 1
```js
const EXPECTED = [
    { suit: 'Hearts', value: 'A' }, { suit: 'Hearts', value: '2' }, { suit: 'Hearts', value: '3' }, { suit: 'Hearts', value: '4' }, { suit: 'Hearts', value: '5' }, { suit: 'Hearts', value: '6' }, { suit: 'Hearts', value: '7' }, { suit: 'Hearts', value: '8' }, { suit: 'Hearts', value: '9' }, { suit: 'Hearts', value: '10' }, { suit: 'Hearts', value: 'J' }, { suit: 'Hearts', value: 'Q' }, { suit: 'Hearts', value: 'K' },
    { suit: 'Diamonds', value: 'A' }, { suit: 'Diamonds', value: '2' }, { suit: 'Diamonds', value: '3' }, { suit: 'Diamonds', value: '4' }, { suit: 'Diamonds', value: '5' }, { suit: 'Diamonds', value: '6' }, { suit: 'Diamonds', value: '7' }, { suit: 'Diamonds', value: '8' }, { suit: 'Diamonds', value: '9' }, { suit: 'Diamonds', value: '10' }, { suit: 'Diamonds', value: 'J' }, { suit: 'Diamonds', value: 'Q' }, { suit: 'Diamonds', value: 'K' },
    { suit: 'Spades', value: 'A' }, { suit: 'Spades', value: '2' }, { suit: 'Spades', value: '3' }, { suit: 'Spades', value: '4' }, { suit: 'Spades', value: '5' }, { suit: 'Spades', value: '6' }, { suit: 'Spades', value: '7' }, { suit: 'Spades', value: '8' }, { suit: 'Spades', value: '9' }, { suit: 'Spades', value: '10' }, { suit: 'Spades', value: 'J' }, { suit: 'Spades', value: 'Q' }, { suit: 'Spades', value: 'K' },
    { suit: 'Clubs', value: 'A' }, { suit: 'Clubs', value: '2' }, { suit: 'Clubs', value: '3' }, { suit: 'Clubs', value: '4' }, { suit: 'Clubs', value: '5' }, { suit: 'Clubs', value: '6' }, { suit: 'Clubs', value: '7' }, { suit: 'Clubs', value: '8' }, { suit: 'Clubs', value: '9' }, { suit: 'Clubs', value: '10' }, { suit: 'Clubs', value: 'J' }, { suit: 'Clubs', value: 'Q' }, { suit: 'Clubs', value: 'K' }
];

describe("cardDeck", () => {
  it("should return the created deck", () => {
    expect(createCardDeck()).toEqual(EXPECTED);
  })
});
```
opção 2
```js
const SUITES = [ 'A', 'B' ];
const VALUES = [1, 2, 3];

describe("cardDeck", () => {
  it("should return the created deck", () => {
    expect(createCardDeck(SUITES, VALUES)).toEqual([
      { suit: 'A', value: 1 }, { suit: 'A', value: 2 }, { suit: 'A', value: 3 }, 
      { suit: 'B', value: 1 }, { suit: 'B', value: 2 }, { suit: 'B', value: 3 }
    ]);
  })
});
